addons/survey/models/survey_user.py

```diff
@@ -422,53 +430,87 @@ def _check_answer_type_skipped(self):
 
     def write(self, vals):
-        score = self._get_answer_score(
-            vals.get('user_input_id'),
-            vals.get('suggested_answer_id'),
-            compute_speed_score=False)
-        if score and not vals.get('answer_score'):
-            vals.update({'answer_score': score})
+        score_vals = self._get_answer_score_values(vals, compute_speed_score=False)
+        if not vals.get('answer_score'):
+            vals.update(score_vals)
         return super(SurveyUserInputLine, self).write(vals)
 
     @api.model
-    def _get_answer_score(self, user_input_id, suggested_answer_id, compute_speed_score=True):
-        """ If score depends on the speed of the answer, we need to compute it.
-        If the user answers in less than 2 seconds, he gets 100% of the points.
-        If he answers after that, he gets minimum 50% of the points.
-        The 50 other % are ponderated between the time limit and the time it took him to answer.
-
-        If the score does not depend on the speed, we just return the answer_score field of the
-        suggested survey.question.answer. """
+    def _get_answer_score_values(self, vals, compute_speed_score=True):
+        """ Get values for: answer_is_correct and associated answer_score.
+
+            Requires vals to contain 'answer_type', 'question_id', and 'user_input_id'.
+            Depending on 'answer_type' additional value of 'suggested_answer_id' may also be
+            required.
+
+            Calculates whether an answer_is_correct and its score based on 'answer_type' and
+            corresponding question. Handles choice (answer_type == 'suggestion') questions
+            separately from other question types. Each selected choice answer is handled as an
+            individual answer.
+
+            If score depends on the speed of the answer, it is adjusted as follows:
+             - If the user answers in less than 2 seconds, they receive 100% of the possible points.
+             - If user answers after that, they receive 50% of the possible points + the remaining
+                50% scaled by the time limit and time taken to answer [i.e. a minimum of 50% of the
+                possible points is given to all correct answers]
+
+            Example of returned values:
+                * {'answer_is_correct': False, 'answer_score': 0} (default)
+                * {'answer_is_correct': True, 'answer_score': 2.0}
+        """
 
-        if not suggested_answer_id:
-            return None
+        answer_type = vals.get('answer_type')
+        question_id = vals.get('question_id')
+        user_input_id = vals.get('user_input_id')
+        question = self.env['survey.question'].browse(int(question_id))
 
-        answer = self.env['survey.question.answer'].search([('id', '=', suggested_answer_id)], limit=1)
-        answer_score = answer.answer_score
+        # default and non-scored questions
+        score_vals = {'answer_is_correct': False}
```

Comment:
I would define a `answer_is_correct` variable here instead and at the end of the method.

```py
return {
    'answer_is_correct': answer_is_correct,
    'answer_score': answer_score
}
```

Reply:
Done!